Implement TTL controller and use the ttl annotation attached to node in secret manager

wojtek-t opened the pull request:
For every secret attached to a pod as volume, Kubelet is trying to refresh it every sync period. Currently Kubelet has a ttl-cache of secrets of its pods and the ttl is set to 1 minute. That means that in large clusters we are targetting (5k nodes, 30pods/node), given that each pod has a secret associated with ServiceAccount from its namespaces, and with large enough number of namespaces (where on each node (almost) every pod is from a different namespace), that resource in ~30 GETs to refresh all secrets every minute from one node, which gives ~2500QPS for GET secrets to apiserver.

Apiserver cannot keep up with it very easily.

Desired solution would be to watch for secret changes, but because of security we don't want a node watching for all secrets, and it is not possible for now to watch only for secrets attached to pods from my node.

So as a temporary solution, we are introducing an annotation that would be a suggestion for kubelet for the TTL of secrets in the cache and a very simple controller that would be setting this annotation based on the cluster size (the large cluster is, the bigger ttl is). 
That workaround mean that only very local changes are needed in Kubelet, we are creating a well separated very simple controller, and once watching "my secrets" will be possible it will be easy to remove it and switch to that. And it will allow us to reach scalability goals.

@dchen1107 @thockin @liggitt

wojtek-t committed:
Implement ttl controller

wojtek-t committed:
Use secret TTL annotation in secret manager

wojtek-t committed:
RBAC for TTL controller

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40796)
<!-- Reviewable:end -->

wojtek-t commented on pkg/kubelet/secret/secret_manager.go:
```diff
@@ -67,6 +69,8 @@ func (s *simpleSecretManager) RegisterPod(pod *v1.Pod) {
 func (s *simpleSecretManager) UnregisterPod(pod *v1.Pod) {
 }
 
+type GetObjectTTLFunc func() (time.Duration, bool)
```
@yujuhong - do you want to review the whole PR, or just the kubelet part and I should find someone else for the controller part?

yujuhong replied:
I'll review everything, but it'd be better if you find someone who's more well-versed in controllers to take a look at that part.

wojtek-t replied:
Ack

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,221 @@
+		}
+		nodeCopy := objCopy.(*v1.Node)
+		setIntAnnotation(nodeCopy, ObjectTTLAnnotation, desiredTTL)
+		_, err = ttlc.kubeClient.Core().Nodes().Update(nodeCopy)
```
Can we patch the node instead to avoid unnecessary conflicts?

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,221 @@
+	node := obj.(*v1.Node)
+	desiredTTL := ttlc.getDesiredTTL()
+
+	if value, ok := getIntFromAnnotation(node, ObjectTTLAnnotation); !ok || value != desiredTTL {
```
(optional) nit: return early to avoid more indentation.

```go
value, ok := getIntFromAnnotation(node, ObjectTTLAnnotation); 
if ok && value == desired {
  return nil
}
```

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,221 @@
+		return
+	}
+
+	workers := 5
```
Why 5 workers?

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -415,7 +415,8 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
 
-	secretManager, err := secret.NewSimpleSecretManager(kubeClient)
+	secretManager, err := secret.NewCachingSecretManager(
+		kubeClient, secret.GetObjectTTLFromNodeFunc(nodeStore, string(nodeName)))
```
Can just use the `kl.GetNode()` function to get the current node object.

yujuhong commented on pkg/kubelet/secret/secret_manager.go:
```diff
@@ -146,6 +153,39 @@ func (s *secretStore) Delete(namespace, name string) {
 }
 
+const (
+	ObjectTTLAnnotation = "node.kubernetes.io/ttl"
```
Why defining this in two places?

wojtek-t replied:
I was wondering where should we put it so that both controller and kubelet can use it (without cyclic dependencies etc.)

wojtek-t replied:
Moved to pkg/api/helpers.go (there were a bunch of other annotations there).

yujuhong commented on pkg/kubelet/secret/secret_manager.go:
```diff
@@ -93,15 +97,18 @@ type secretStore struct {
 		kubeClient: kubeClient,
 		clock:      clock,
 		items:      make(map[objectKey]*secretStoreItem),
 		ttl:        ttl,
```
In this case, I don't think we need a `ttl` field anymore. Just define a default constant in this file is enough.

yujuhong commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:
```diff
@@ -907,5 +907,28 @@ items:
     - create
     - patch
     - update
+- apiVersion: rbac.authorization.k8s.io/v1beta1
```
lgtm, but add @liggitt just in case.

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,221 @@
+limitations under the License.
+*/
+
+package ttl
```
I think the controller needs more logging/events to surface the fact that it's changing the TTL.

fgrzadkowski replied:
+1

wojtek-t replied:
I added logs. Let me know if these should be events instead (or additionally).

wojtek-t reviewed (commented):
@yujuhong - thanks for comments. I applied or answered them. PTAL

wojtek-t commented:
@fgrzadkowski agreed to be the second reviewer of controller part.

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,238 @@
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	"k8s.io/kubernetes/pkg/client/legacylisters"
```
prefer the generated listers.  `pkg/client/listers` I think

wojtek-t replied:
I tried - but it would require some significant refactoring IIUC, because from InformerFactory.Nodes(), there is no easy way to create this generated lister...

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,238 @@
+		return
+	}
+
+	workers := 5
```
I think many/most take it as a parameter and we set it to five.

wojtek-t replied:
changed to param

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func (ttlc *TTLController) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting TTL controller")
```
add `defer utilruntime.HandleCrash()` so a mistake here doesn't crash the process

deads2k replied:
defer shutting down your work queue.

wojtek-t replied:
done

wojtek-t replied:
done

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,238 @@
+	ttlc.queue.Add(key)
+}
+
+func (ttlc *TTLController) worker() {
```
daemonset, replicaset, deployment, and other split this into work and process to avoid the anonymous func.  We're trying to standardize on that structure.  https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md

wojtek-t replied:
done

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	listers "k8s.io/kubernetes/pkg/client/legacylisters"
```
Try to use the generated listers here: https://github.com/kubernetes/kubernetes/tree/master/pkg/client/listers We're actively trying to remove legacylisters.

wojtek-t replied:
I tried doing this. But there doesn't seem to be a way from go from:
"informers.NodeInformer"
(which is what we have in cmd/kuber-controller-manager/app/controllermanager.go"
to
the listers you want me to use.

I.e. from the NodeInformer all I can get is "cache.SharedIndexInformer" from which I don't know how to create the "NodeLister" from the generated lister...

wojtek-t replied:
Ohh wait - maybe it is possible. I thing I missed the GetIndexer() function which is what I need.

deads2k replied:
> Ohh wait - maybe it is possible. I thing I missed the GetIndexer() function which is what I need.

Ok, we you can't do it now, then just an issue to fix up after https://github.com/kubernetes/kubernetes/pull/40385 wil do.

wojtek-t replied:
Actually, I managed to fix it.

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+func (ttlc *TTLController) worker() {
+	for {
+		if quit := ttlc.processItem(); quit {
+			glog.Infof("TTL controller worker shutting down")
```
Is this important enough to log?  The overall shutdown is already logged, a panic wouldn't stop here, so this is extraneous, rigth?

wojtek-t replied:
removed

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+func (ttlc *TTLController) processItem() bool {
+	key, quit := ttlc.queue.Get()
+	if quit {
+		return true
```
others return "should I continue" instead of "should I quit"

wojtek-t replied:
changed (though not all others, e.g. RC is doing what I had before).

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+	if err != nil {
+		return err
+	}
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, &v1.Node{})
```
annotations don't require strategic patches, just json patches, right?  I'm wary of strategic patches now that we're making code more generic.

wojtek-t replied:
Why? Can you clarify?

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+	if ok && value == desiredTTL {
+		return nil
+	}
+	objCopy, err := api.Scheme.DeepCopy(node)
```
yay! :)

wojtek-t replied:
ack

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+}
+
+func (ttlc *TTLController) updateNodeIfNeeded(key string) error {
+	obj, exists, err := ttlc.nodeStore.GetByKey(key)
```
yeah, this gets cleaner with the new generated listers.  It matches the client semantic of errors.IsNotFound

wojtek-t replied:
done

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+func (ttlc *TTLController) updateNode(_, newObj interface{}) {
+	node, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("Unexpected object type: %v", newObj)
```
`utilruntime.Error` here and elsewhere this is a severe problem.

wojtek-t replied:
done

deads2k commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,241 @@
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down TTL controller")
```
defering this would pair nicely.  What do you think?

wojtek-t replied:
done

deads2k commented:
Simple, tight controller.  Permissions look good.  I had some minor comments.

fgrzadkowski commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+	case nodeCount <= 1000:
+		return 30
+	case nodeCount <= 2000:
+		return 60
```
I thought that any value below 60s does not make much sense. Did I miss something?

wojtek-t replied:
Thanks for clarifying @yujuhong - that's exactly why I did this.

fgrzadkowski commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+		utilruntime.HandleError(fmt.Errorf("unexpected object types: %v", newObj))
+		return
+	}
+	ttlc.enqueueNode(node)
```
Why do you enqueue with node update? I think this means that will process every node at least twice - first when it's added and later when see our own modification. Also - is it going to be triggered when node updates conditions/status?

wojtek-t replied:
Yes - this is on purpose. So process each node periodically and ensure that all of them have the same ttl. See my comment above also.

fgrzadkowski commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+		utilruntime.HandleError(fmt.Errorf("unexpected object types: %v", obj))
+		return
+	}
+	atomic.AddInt32(&ttlc.nodeCount, 1)
```
I think that we will set different TTL for each node. That would be confusing... If it's hard to fix then we should at least document it properly.

yujuhong replied:
Another question is that IIUC, you'll only be changing the TTL of new nodes (or nodes that are updated). If we scale a cluster from 1 to 5000 nodes gradually, the first 100 nodes will have 0s TTL, while the last 3000 will have 300s TTL. That doesn't seem desirable. Can we use the size of the cluster as a trigger to update the annotations for all nodes in the cluster instead?

wojtek-t replied:
That's not true. There might be some periods where some nodes have different ttls than other nodes, but those should be very short periods.
Note that Kubelet is updating status every 10 seconds. As a result, we will be processing each node every 10s. That means, that once the cluster changes the size to from one group to the other, within 10s or so, all nodes should have updated ttls.
I think this is good enough.

I'm adding a comment to clarify this.

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+		return
+	}
+	atomic.AddInt32(&ttlc.nodeCount, -1)
+	ttlc.enqueueNode(node)
```
I don't think we need to process the deleted node. In fact, `ttlc.updateNodeIfNeeded(key.(string))` in line 154 will return error, forcing you to requeue the item again.

wojtek-t replied:
Good point. Fixed.

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+func (ttlc *TTLController) updateNodeIfNeeded(key string) error {
+	node, err := ttlc.nodeStore.Get(key)
+	if err != nil {
+		return err
```
We should distinguish the node "not found" error and return nil.

wojtek-t replied:
Done.

yujuhong commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -281,6 +281,7 @@ func newControllerInitializers() map[string]InitFunc {
 	controllers["statefuleset"] = startStatefulSetController
 	controllers["cronjob"] = startCronJobController
 	controllers["certificatesigningrequests"] = startCSRController
+	controllers["ttl"] = startTTLController
```
I had the same thought, but I am pretty bad at naming.

wojtek-t replied:
I'm fine with changing but i couldn't come up with any better name - so if you could suggest something better, I'm happy to change that.

fgrzadkowski replied:
I can't find anything better. Let's leave it as it is :/

liggitt commented:
Sorry, just made it to this. Implementing dynamic kubelet config one controller-imposed annotation at a time seems like a bad direction to me. Can we not make it part of the kubelet config, kubemark can tune it appropriately, and when we get dynamic config working we'll get this dynamic updating for free?

smarterclayton commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+const (
+	// ObjectTTLAnnotation is a suggestion for kubelet for how long it can cache
+	// an object (e.g. secret, config map) before fetching from apiserver again.
+	ObjectTTLAnnotation = "node.kubernetes.io/ttl"
```
All new annotations that are API should be alpha.

wojtek-t replied:
Done

smarterclayton commented:
Agree that a more limited solution that works in a scope is better than adding a brand new top level item.  Dynamic kubelet config should cover this feature, not the other way around.

smarterclayton commented:
If this isn't covered under the dynamic kubelet config design proposal - why not?

wojtek-t commented:
@smarterclayton @liggitt - let me clarify this. Improving how Kubelet is fetching secrets is necessary to improve scalability of the system (e.g. scale to 5000-node clusters). Since we would like to do it for 1.6, dynamic kubelet config is not going to meet this.

That said, this PR is just a temporary solution. In the medium term, Kubelet should just be watching "all secrets that it needs (i.e. that are attached to its pods)" (this is security requirement, right) Once this will be possible to do in apiserver, we will just get rid of this controller at all. We will try to prioritize the work of enabling doing this in apiserver for 1.7. 

That said, I'm explicit comment about this to the file.

wojtek-t commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,237 @@
+	// nodeStore is a local cache of nodes.
+	nodeStore listers.NodeLister
+
+	// Secrets that need to be synced.
```
fixed

wojtek-t commented on pkg/kubelet/secret/secret_manager.go:
```diff
@@ -31,6 +31,8 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/client-go/util/clock"
+
```
Removed (issue debugged in the previous PR: #40795 )

wojtek-t commented:
@smarterclayton @liggitt - and thing that is probably worth mentioning is that this approach was discussed and approved by @thockin and @dchen1107

liggitt commented:
> Improving how Kubelet is fetching secrets is necessary to improve scalability of the system (e.g. scale to 5000-node clusters). Since we would like to do it for 1.6, dynamic kubelet config is not going to meet this.

Making the TTL a config option now makes tuning for the "5000 node cluster" scenario possible in 1.6, and fits into the dynamic config design well. In the short-term, it means that the tuning has to be applied from the beginning (instead of being able to tune original nodes longer as the number of nodes grows), but it doesn't seem unreasonable to expect a cluster admin to know what order of magnitude they expect their cluster to be ("do you expect 5 nodes? 50? 500? 5000?)

> That said, this PR is just a temporary solution. In the medium term, Kubelet should just be watching "all secrets that it needs (i.e. that are attached to its pods)" (this is security requirement, right) Once this will be possible to do in apiserver, we will just get rid of this controller at all. We will try to prioritize the work of enabling doing this in apiserver for 1.7.

This seems like a really heavyweight temporary solution. It's quite possible we'll get dynamic config before node-specific secret watching.

wojtek-t commented:
> This seems like a really heavyweight temporary solution. It's quite possible we'll get dynamic config before node-specific secret watching.

I completely disagree with that it is heavy. It is a single PR written within few hours. This definitely wouldn't be the case for dynamic kubelet config. And medium/long term we want to have it watch-based anyway. 

> but it doesn't seem unreasonable to expect a cluster admin to know what order of magnitude they expect their cluster to be ("do you expect 5 nodes? 50? 500? 5000?)

I don't really agree. Most of people start with some small cluster for the testing purposes and then it is growing. It can be hard to say up-front what size it would have.

That said, I definitely think that the medium/long term solution should be watch based and once this is done, we don't need any TTL at all. That said, if this is supposed to be a temporary solution, we should make it as small and as independent as possible. And this solution satisfies those requirements:
- there are pretty much 20 lines of changes in Kubelet (+ tests)
- and a completely separate controller which will be super easy to remove once we no longer need it.

liggitt commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,291 @@
+	<-stopCh
+}
+
+func (ttlc *TTLController) addNode(obj interface{}) {
```
doesn't this get re-fired on resync?

wojtek-t replied:
the update() function is supposed to be called on resync.

liggitt commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+func (ttlc *TTLController) deleteNode(obj interface{}) {
+	_, ok := obj.(*v1.Node)
```
can we get a DeletedFinalStateUnknown object here?

wojtek-t replied:
good point. Fixed.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -265,6 +265,13 @@ func init() {
 	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "ttl-controller"},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("update", "patch").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
```
since this logically needs list/watch on nodes, include it in the role (even though it's using a shared informer)

wojtek-t replied:
Sure. Fixed.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -265,6 +265,13 @@ func init() {
 		},
 	})
 	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "ttl-controller"},
```
`ttl-controller` is pretty vague... trying to think of a name that captures what the ttl is for

wojtek-t replied:
Yeah, though I thought about it and didn't have any better name. But if you have some suggestion, I'm happy to change that.

liggitt commented:
I still think a controller for this is overkill, but I'm willing to be overruled

liggitt commented on pkg/kubelet/secret/secret_manager.go:
```diff
@@ -149,6 +159,31 @@ func (s *secretStore) Delete(namespace, name string) {
+		}
+		if node != nil && node.Annotations != nil {
+			if value, ok := node.Annotations[v1.ObjectTTLAnnotationKey]; ok {
+				if intValue, err := strconv.Atoi(value); err == nil {
```
make sure int is positive

wojtek-t commented:
@yujuhong @fgrzadkowski - the PR should be ready for final review. I guess the only outstanding comment is about the name (which I'm happy to change, but don't have a better suggestion). 
PTAL

wojtek-t commented:
@k8s-bot gci gke test this, issue: #IGNORE (up failed)

smarterclayton commented:
> and thing that is probably worth mentioning is that this approach was discussed and approved by @thockin and @dchen1107

That's not sufficient justification to not include a high level description of what this is intended to achieve.  There are lots of other stakeholders in Kube who need to understand why things are changing across many different teams and sigs, and if you don't write a proposal I expect at least a summary in the PR.  Controllers add operational complexity and overhead, performance impact, cause new and surprising behaviors.

Please update the description of the PR with a quick high level summary that summarizes the tradeoffs here so that others beside the mentioned folks can understand and comment about the intent.

yujuhong commented on pkg/kubelet/secret/secret_manager.go:
```diff
@@ -93,15 +100,18 @@ type secretStore struct {
 	items map[objectKey]*secretStoreItem
-	ttl   time.Duration
+
+	ttl    time.Duration
```
nit: rename this as `defaultTTL` to be precise.

yujuhong commented on pkg/controller/ttl/ttlcontroller_test.go:
```diff
@@ -0,0 +1,166 @@
+		actions := fakeClient.Actions()
+		assert.Equal(t, 1, len(actions), "unexpected actions: %#v", actions)
+		patchAction := actions[0].(core.PatchActionImpl)
+		if testCase.patch != string(patchAction.Patch) {
```
nit: `assert.Equal(t, testCase.Patch, string(patchAction.Patch), "test case [%d]", i)`

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,299 @@
+		glog.V(2).Infof("Failed to change ttl annotation for node %s: %v", node.Name, err)
+		return err
+	}
+	glog.V(2).Infof("Changed ttl annotation for node %s: %v", node.Name, err)
```
No need to print `err` since it's nil. Printing the actual TTL would be more useful.

wojtek-t replied:
thanks for catching (that's what it was supposed to be)

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,299 @@
+	// has to be flapping constantly betwenn <=X-1 and X+1<=.
+	// WARNING: you can't make this "1" a higher constant, as it may result
+	// in different nodes having different ttls over long period of time, e.g.
+	// if we swap the 1 with say 5, then in the scenario:
```
Not sure why this is a problem. You can modify the internal, desired TTL whenever the node count changes. All nodes can just use this internal TTL (instead of calculating it individually). Basically, only change the TTL if the N (the node count) is greater than Tmax or fall below Tmin, where Tmax - Tmin > M. M can be a buffer that's larger than 1.

Having a buffer of only 1 (the current implementation) doesn't seem safe, especially in a 5000 node cluster.

wojtek-t replied:
I thought about it and i though that it wouldn't work. But you're right, this actually would work ;)
The only problem with that is restart, because after restart we may potentially re-annotate all nodes. But that shouldn't be a big deal.
Changing.

yujuhong commented:
Applied "do-not-merge" based on @smarterclayton's comment: https://github.com/kubernetes/kubernetes/pull/40796#issuecomment-277815718

wojtek-t commented:
> Please update the description of the PR with a quick high level summary that summarizes the tradeoffs here so that others beside the mentioned folks can understand and comment about the intent.

@smarterclayton - Sorry. To be honest, I was convinced that I added a PR description initially. But it seems I didn't (I only described it in the previous PR (which we closed) and forgot to copy it into PR description).
I obviously agree that we need transparency and it should be clear what, how and why we are doing. Hopefully the description I've just added is making it clear. Sorry for not doing it while opening the PR.
PTAL

wojtek-t reviewed (commented):
@yujuhong - PTAL

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,297 @@
+	// Returns true if all underlying informers are synced.
+	hasSynced func() bool
+
+	lock sync.Mutex
```
Can probably use `sync.RWMutex` since most of the time you'll have multiple readers, but no writers.

wojtek-t replied:
done

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,297 @@
+		glog.V(2).Infof("Failed to change ttl annotation for node %s: %v", node.Name, err)
+		return err
+	}
+	glog.V(2).Infof("Changed ttl annotation for node %s to %d", node.Name, value)
```
nit: s/%d/%d seconds

wojtek-t replied:
done

yujuhong commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,297 @@
+	// Processing all updates of nodes guarantees that we will update
+	// the ttl annotation, when cluster size changes.
+	// We are relying on the fact that Kubelet is updating node status
+	// every 10s (or generaly every X seconds), which means that whenever
```
nit: s/generaly/generally

wojtek-t replied:
done

fgrzadkowski commented on pkg/controller/ttl/ttlcontroller.go:
```diff
@@ -0,0 +1,297 @@
+		ttlc.lock.Lock()
+		defer ttlc.lock.Unlock()
+		ttlc.nodeCount++
+		if ttlc.nodeCount > ttlBoundaries[ttlc.boundaryStep].sizeMax {
```
This will break if someone has more than 10000 nodes in a cluster.

wojtek-t replied:
Added one more step to MaxInt32 - that should be enough :)

fgrzadkowski reviewed (commented):
Only 1 small comment. In general LGTM.

I have only reviewed the controller part.

smarterclayton commented:
Ok, that addresses my concern for now.  I would also prefer a server side watch that hatches secret changes on a minimum burst interval - that overlaps with some of the changes discussed for efficient bulk namespace watching and access control of watches (losing access terminates the watch).  I'll queue that in those discussions.

smarterclayton commented:
"Batches"

wojtek-t commented:
> Ok, that addresses my concern for now. I would also prefer a server side watch that hatches secret changes on a minimum burst interval - that overlaps with some of the changes discussed for efficient bulk namespace watching and access control of watches (losing access terminates the watch). I'll queue that in those discussions.

@smarterclayton - yes that makes perfect sense to me. Is there already an issue where this is discussed (or those were just f2f discussions)?

wojtek-t reviewed (commented):
Comments applied.

Since @smarterclayton concerns are addressed by the PR description, I'm removing "do-not-merge" label.

@yujuhong @fgrzadkowski - if you're fine, can you add "lgtm" label?

wojtek-t commented:
@k8s-bot kops test this, issue: #IGNORE

wojtek-t commented:
@k8s-bot gce etcd3 test this, issue: #36379

yujuhong commented:
@wojtek-t looks good overall, but I am concerned about the lack of e2e/integration tests. Is it possible to add some?

wojtek-t committed:
Add integration test for ttlcontroller.

wojtek-t commented:
> @wojtek-t looks good overall, but I am concerned about the lack of e2e/integration tests. Is it possible to add some?

For e2e test it would be tricky, as it would require signifcantly changing size of the cluster.
But integration test sounds fine for it - I will try to add one.

wojtek-t commented:
@yujuhong - I added integration test for it. PTAL

fgrzadkowski commented:
I reviewed integration test and it LGTM. Since earlier I reviewed controller part and @yujuhong reviewed kubelet part and the only remaining concerns were tests I'm adding LGTM label.

fgrzadkowski commented:
/lgtm
